File: src/main/java/fr/lewon/bot/runner/bot/AbstractBotBuilder.java
package fr.lewon.bot.runner.bot;

import fr.lewon.bot.runner.bot.operation.BotOperation;
import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor;
import fr.lewon.bot.runner.bot.props.BotPropertyStore;
import fr.lewon.bot.runner.bot.task.BotTask;
import fr.lewon.bot.runner.errors.InvalidBotPropertyValueException;
import fr.lewon.bot.runner.errors.MissingBotPropertyException;
import fr.lewon.bot.runner.util.BeanUtil;
import fr.lewon.bot.runner.util.BotPropertyParser;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.Map;

public abstract class AbstractBotBuilder {

    private static BotPropertyParser botPropertyParser = BeanUtil.getBean(BotPropertyParser.class);

    private String botName;
    private List<BotPropertyDescriptor> botPropertyDescriptors;

    public AbstractBotBuilder(String botName, BotPropertyDescriptor... botPropertyDescriptors) {
        this(botName, Arrays.asList(botPropertyDescriptors));
    }

    public AbstractBotBuilder(String botName, List<BotPropertyDescriptor> botPropertyDescriptors) {
        this.botName = botName;
        this.botPropertyDescriptors = Collections.unmodifiableList(botPropertyDescriptors);
    }

    public Bot buildBot(Map<String, String> properties) throws InvalidBotPropertyValueException, MissingBotPropertyException {
        BotPropertyStore botPropertyStore = botPropertyParser.parseParams(properties, this.botPropertyDescriptors);
        return new Bot(botPropertyStore, (b) -> this.getInitialTasks(b), this.getBotOperations());
    }

    protected abstract List<BotOperation> getBotOperations();

    protected abstract List<BotTask> getInitialTasks(Bot bot);

}

File: src/main/kotlin/fr/lewon/bot/runner/bot/AbstractBotBuilder.kt
package fr.lewon.bot.runner.bot

import fr.lewon.bot.runner.bot.operation.BotOperation
import fr.lewon.bot.runner.bot.props.BotPropertyDescriptor
import fr.lewon.bot.runner.bot.task.BotTask
import fr.lewon.bot.runner.errors.InvalidBotPropertyValueException
import fr.lewon.bot.runner.errors.MissingBotPropertyException
import fr.lewon.bot.runner.util.BeanUtil
import fr.lewon.bot.runner.util.BotPropertyParser
import org.springframework.web.reactive.function.client.WebClient

abstract class AbstractBotBuilder(private val botName: String, private val botPropertyDescriptors: List<BotPropertyDescriptor>) {

    constructor(botName: String, vararg botPropertyDescriptors: BotPropertyDescriptor) : this(botName, listOf(*botPropertyDescriptors)) {}

    @Throws(InvalidBotPropertyValueException::class, MissingBotPropertyException::class)
    fun buildBot(properties: Map<String, String?>): Bot {
        val botPropertyStore = botPropertyParser.parseParams(properties, this.botPropertyDescriptors)
        return Bot(botPropertyStore, { b -> this.getInitialTasks(b) }, this.getBotOperations(), buildWebClient())
    }

    protected abstract fun getBotOperations(): List<BotOperation>

    protected abstract fun buildWebClient(): WebClient

    protected abstract fun getInitialTasks(bot: Bot): List<BotTask>

    companion object {
        private val botPropertyParser = BeanUtil.getBean(BotPropertyParser::class.java)
    }

}
